pilot/pkg/xds/fake.go

```diff
@@ -0,0 +1,330 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"bytes"
+	"reflect"
+	"text/template"
+
+	"github.com/Masterminds/sprig"
+	endpoint "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	listener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	hcm "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	envoy_extensions_filters_network_tcp_proxy_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/tcp_proxy/v3"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/wellknown"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/any"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+
+	cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	route "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+
+	"istio.io/istio/pilot/pkg/config/kube/crd"
+	"istio.io/istio/pilot/pkg/config/memory"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/plugin"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pilot/pkg/serviceregistry/serviceentry"
+	"istio.io/istio/pkg/config/mesh"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/test"
+)
+
+type FakeOptions struct {
+	// If provided, these configs will be used directly
+	Configs []model.Config
+	// If provided, the yaml string will be parsed and used as configs
+	ConfigString string
+	// If provided, the ConfigString will be treated as a go template, with this as input params
+	ConfigTemplateInput interface{}
+	// If provided, this mesh config will be used
+	MeshConfig   *meshconfig.MeshConfig
+	MeshNetworks *meshconfig.MeshNetworks
+}
+
+type FakeDiscoveryServer struct {
+	t           test.Failer
+	Store       model.ConfigStore
+	Discovery   *DiscoveryServer
+	PushContext *model.PushContext
+	Env         *model.Environment
+}
+
+func getConfigs(t test.Failer, opts FakeOptions) []model.Config {
```

Comment:
I don't think you want to have production code dependent on go's test framework. Maybe this should be under `pkg/test` somewhere?

Reply:
We have a test that ensures the `testing` flags are not linked into the binaries, and this seems to pass it, suggesting to me its not actually linked in? Not an expert on how go handles this though, and I would be fine to move it for other reasons as well. WDYT?